File: exercism/collatz_conjecture.ml
let collatz_conjecture n =
  let rec helper num cnt =
    if num <= 0 then Error "Only positive integers are allowed"
    else
      match num with
      | 1 -> Ok cnt
      | a ->
          if a mod 2 = 0 then helper (a / 2) (cnt + 1)
          else helper ((3 * a) + 1) (cnt + 1)
  in
  helper n 0

File: exercism/collatz_conjecture.cpp
#include "collatz_conjecture.h"
#include <stdexcept>

namespace collatz_conjecture {
    int helper(int n, int cnt) {
        if (1 == n) {
            return cnt;
        } else {
            if (0 == n % 2) {
                return helper(n / 2, cnt + 1);
            } else {
                return helper(3 * n + 1, cnt + 1);
            }
        }
    }

    int steps(int n) {
        if (n <= 0) {
            throw std::domain_error("Only positive integers are allowed");
        }
        return helper(n, 0);
    }
}  // namespace collatz_conjecture
